network/tcp_client.cc

```diff
@@ -0,0 +1,490 @@
+        hs_response->GetMinorVersion() != hs_request.GetMinorVersion())
+    {
+        return absl::FailedPreconditionError(
+        absl::StrCat("PerformHandshake: Handshake version mismatch. Server is v",
+                     hs_response->GetMajorVersion(),
+                     ".",
+                     hs_response->GetMinorVersion(),
+                     " Client requires v",
+                     hs_request.GetMajorVersion(),
+                     ".",
+                     hs_request.GetMinorVersion()));
+    }
+    std::cout << "Client: Handshake versions compatible." << std::endl;
+    return absl::OkStatus();
+}
+
+absl::Status TcpClient::StartKeepAlive()
+{
+    if (m_keep_alive.running.load())
+    {
+        return absl::AlreadyExistsError("StartKeepAlive: Keep-alive is already running.");
+    }
+    if (!IsConnected())
+    {
+        return absl::FailedPreconditionError("StartKeepAlive: Client is not connected.");
+    }
+
+    m_keep_alive.running.store(true);
+    m_keep_alive.thread = std::thread(&TcpClient::KeepAliveLoop, this);
+
+    std::cout << "Client: Keep-alive thread started. Interval: " << m_keep_alive.interval_sec
+              << " s, Ping Timeout: " << m_keep_alive.ping_timeout_ms << " ms." << std::endl;
+    return absl::OkStatus();
+}
+
+void TcpClient::KeepAliveLoop()
+{
+    while (m_keep_alive.running.load())
+    {
+        std::unique_lock<std::mutex> lk(m_keep_alive.mutex);
+        if (m_keep_alive.cv.wait_for(lk, std::chrono::seconds(m_keep_alive.interval_sec), [this] {
+                return !m_keep_alive.running.load();
+            }))
+        {
+            std::cout << "KeepAliveLoop: Client Keep-Alive Thread: Stop signaled via CV."
+                      << std::endl;
+            break;
+        }
+        lk.unlock();
+
+        if (!m_keep_alive.running.load())
+        {
+            break;
+        }
+
+        if (IsConnected())
+        {
+            auto ping_status = PingServer(m_keep_alive.ping_timeout_ms);
+            if (!ping_status.ok())
+            {
+                std::cout << "KeepAliveLoop: Ping failed. Reason: " << ping_status.message()
+                          << std::endl;
+                SetClientStatus(ClientStatus::STATUS_ERROR);
+                m_keep_alive.running.store(false);
+            }
+        }
+        else
+        {
+            std::cout << "KeepAliveLoop: Client not connected. Stopping keep-alive." << std::endl;
+            m_keep_alive.running.store(false);
+        }
+    }
+}
+
+void TcpClient::StopKeepAlive()
+{
+    if (!m_keep_alive.running.exchange(false))
+    {
+        return;
+    }
+    m_keep_alive.cv.notify_one();
+    if (m_keep_alive.thread.joinable())
+    {
+        m_keep_alive.thread.join();
+    }
+}
+
+ClientStatus TcpClient::GetClientStatus() const
+{
+    std::lock_guard<std::mutex> lock(m_status_mutex);
+    return m_status;
+}
+
+ClientStatus TcpClient::SetClientStatus(ClientStatus status)
```

Comment:
when do we need the return value? and why do we return `return m_status = status;`